File: cPP_java_py_js-DSA/1ARRAY_STRINGS/remove-duplicate-sortedArray/removeDuplicateSortedArray.js
import { printComplexityOfSolution } from "../../commonFunctions/printComplexities.js";
// Problem : Given sorted array and we need remove all the duplicate elements and left behind only the non-duplicate element in the array
/*------------------------------------------------------------ */

/*---Optimised Approach - StartIndex pointer and comparison--------------------------------------------------------- */

function removeDuplicatedSortedArray_Approach1_Optimised(arr) {
  // concept : Two pointer for start index - In-place
  // time :o(n), space :O(1)
  printComplexityOfSolution(
    "Optimised- two pointer comparison",
    "O(N)",
    "O(1)",
  );
  let n = arr.length;

  let startIndex = 0;
  let count = 1;
  for (let i = 1; i < arr.length; i++) {
    // step1 : find place the unique values at right position
    if (arr[i] != arr[startIndex]) {
      startIndex++;
      arr[startIndex] = arr[i];
      count++; // count of unique values
    }
  }

  // step 2: remove the extra elements other than unique from existing array
  for (let i = startIndex + 1; i < n; i++) {
    arr.pop();
  }
  console.log("After Removing duplicate elements ", arr);
  return count;
}
let arr = [1, 1, 3, 3, 12, 22, 42];
console.log(arr);
console.log(removeDuplicatedSortedArray_Approach1_Optimised(arr));

File: cPP_java_py_js-DSA/1ARRAY_STRINGS/remove-duplicate-sortedArray/remove-duplicate-sorted-array.cpp
#include <bits/stdc++.h>
#include "../../0commonFunctions/printComplexity.h"
using namespace std;

// Remove duplicate from sorted array  : Given an array of integers and we need to remove the duplicate elements from the sorted array

vector<int> removeDuplicatedFromSortedArray_Approach_Brute(vector<int> arr)
{

    printComplexities("O(n)-> array traversal  + O(N log n)->sorting ", "O(n)-> hashing map space");
    // concept: brute force : Map based , time : O(n log n )  , space : "O(n)"

    // Step1 : update the  frequency of elements
    unordered_map<int, bool> mp;
    vector<int> ans;
    for (auto val : arr)
    {
        mp[val] = true;
    }
    // Step2 :Iterate over map and  store the keys in ans array
    for (auto it : mp)
    {
        ans.push_back(it.first);
    }

    // sort the ans array and return it
    sort(ans.begin(), ans.end());

    return ans;
}

int main()
{
    cout << "Remove the Duplicates from sorted array " << endl;
    vector<int> arr = {0,0,1,1,1,2,2,3,3,4};
    printArray(arr);
    vector<int> ans = removeDuplicatedFromSortedArray_Approach_Brute(arr);
    printArray(ans);
    return 0;
}